src/main/java/bio/overture/ego/security/OAuth2SsoFilter.java

```diff
@@ -61,6 +69,53 @@ public void onAuthenticationSuccess(
   private OrcidService orcidService;
   private GithubService githubService;
 
+  private AuthenticationFailureHandler failureHandler =
+      (request, response, exception) -> {
+        val rootExceptionThrowable = exception.getCause();
+        val application =
+            applicationService.getByClientId(
+                (String) request.getSession().getAttribute("ego_client_id"));
+        String errorRedirectUri =
+            (String) request.getSession().getAttribute("ego_error_redirect_uri");
+
+        val errorRedirect = Redirects.getErrorRedirectUri(application, errorRedirectUri);
+        URIBuilder errorUri;
+
+        if (rootExceptionThrowable instanceof NoPrimaryEmailException) {
+          try {
+            val reqUri = new ArrayList<String>(asList(request.getRequestURI().split("/")));
+            val provider = reqUri.get(reqUri.size() - 1).toUpperCase();
+            errorUri = new URIBuilder(errorRedirect);
+            errorUri.addParameter("error_code", "403");
+            errorUri.addParameter("error_type", rootExceptionThrowable.getMessage());
+            try {
+              ProviderType.resolveProviderType(provider);
+              errorUri.addParameter("provider_type", provider);
+            } catch (IllegalArgumentException e) {
+              log.warn(format("Invalid provider: '%s'", provider));
+            }
+            response.setStatus(403);
+            response.sendRedirect(errorUri.build().toString());
+          } catch (URISyntaxException e) {
+            log.error("Invalid redirect uri");
+          }
+        } else if (rootExceptionThrowable instanceof UserDeniedAuthorizationException) {
+          // User denies Ego access/cancels login
+          try {
+            errorUri = new URIBuilder(errorRedirect);
+            errorUri.addParameter("error_code", "403");
+            // setting explicitly to "access_denied" because LinkedIn error_code differs from other providers
```

Comment:
This is also a good comment but could use more detail about what each provider expects. Do we need seperate responses for linked in vs. other providers?

Reply:
Just tested this with linkedin now that it's working, looks like they come back with an `OAuth2Exception` with an `error="invalid_request"` when the user denies access...so i will need to handle that separately.

Reply:
Actually I can handle both those errors with this block. It might be useful to check the `error_code` too to make sure I'm handling the correct OAuth2Exception scenario

Reply:
Ooh even better: `Condition 'rootExceptionThrowable instanceof UserDeniedAuthorizationException' covered by subsequent condition 'rootExceptionThrowable instanceof OAuth2Exception'`